Correct the callback for extinctions

tpoisot opened the pull request:
@evadelmas  I'm pretty sure these changes correct the behavior of the callback for extinctions (and also make the code simpler in `simulate`) -- I have no negative biomasses on the few simulations I ran.

tpoisot committed:
:hand: attempt to change the callback

tpoisot committed:
:bug:

tpoisot committed:
:hand:

tpoisot committed:
:hand:

tpoisot committed:
:hand: discrete callback

tpoisot committed:
:hand: clean simulate

tpoisot committed:
:hand: correct callback name

tpoisot committed:
:bug: typo

tpoisot committed:
:hand: tests

tpoisot committed:
:wrench: do not save positions after extinction callback

tpoisot committed:
:hand:

tpoisot committed:
:hand:

tpoisot committed:
:hand:

tpoisot committed:
:hand:

tpoisot committed:
:hand: try to force dtmin to true

tpoisot committed:
:hand:

ChrisRackauckas commented on src/simulate.jl:
```diff
@@ -67,9 +67,8 @@ function simulate(parameters, biomass; concentration::Vector{Float64}=rand(Float
-  sol = solve(prob, alg, callback = extinction_callback, saveat=t_keep, dense=false, save_timeseries=false)
-
+  extinction_callback = DiscreteCallback(species_under_extinction_threshold, affect_function; save_positions=(true,true))
+  sol = solve(prob, alg, callback = CallbackSet(extinction_callback, PositiveDomain()), saveat=t_keep, dense=false, save_timeseries=false, force_dtmin=true)
```
`force_dtmin=true` is bad. If you're hitting it often then something is wrong in the model.

tpoisot replied:
I'm starting to get the sense that there are *many* things wrong with the model, but that's the best food web dynamics model we have.

ChrisRackauckas replied:
Why do you need it? What's going on that's causing it? What's going to infinity or diving by zero or ? Why is it not better to just exit if such an error occurs?

tpoisot replied:
When it's removed, the `PositiveDomain()` callback complains (and the solver terminates). It seems to make very little difference in the output. What is weird is that despite the `PositiveDomain` callback, some abundances go below 0.

ChrisRackauckas replied:
Abundences can go below zero by the amount of `abstol`. That's a Shampine thing. But your model is okay with continuing after a dtmin step? It doesn't end up taking billions of small steps? I wonder if this is https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/issues/273

ChrisRackauckas replied:
Ahh, there could be some issue in the callback since I got this issue I forgot about: https://github.com/JuliaDiffEq/DiffEqCallbacks.jl/issues/44

tpoisot replied:
Thank for the clarification about abstol. So the "correct" way to set these species to 0 would be to use `abstol` [here](https://github.com/PoisotLab/BioEnergeticFoodWebs.jl/blob/next-speed/src/simulate.jl#L54)?

tpoisot replied:
The strange thing (from my point of view) is that the model warns about being unable to restrict values to domain maybe once or twice for 500 timesteps, so this seems to be really rare.

ChrisRackauckas replied:
For now `isoutofdomain=(u,p,t)->any(x->x<0,u)` is the safest way to ensure positivity until I find out why that dtmin issue is popping up. With that, since it cannot hit zero, I would keep the extinction check above zero. `10eps()` is essentially zero though. I would take something like `1000eps()` or `1e5eps()` as a threshold to zero since I'm sure most people aren't solving this equation with like `abstol=reltol=1e-12`.

tpoisot replied:
That's a fair assumption. I think the original papers from 2005-ish used 1e-10, but no one is sure because the code got lost. I'll try this later today and see whether that works. Thanks!

tpoisot commented:
Not quite true -- there are situations in which some species reappear.

![extinctions](https://user-images.githubusercontent.com/579329/42892830-cbf99690-8a80-11e8-8631-8dcd6bd5b8e2.png)

@ChrisRackauckas  -- I changed the callback to a discrete one (see [here](https://github.com/PoisotLab/BioEnergeticFoodWebs.jl/blob/next-speed/src/simulate.jl#L53), and removed the line to correct the derivative [here](https://github.com/PoisotLab/BioEnergeticFoodWebs.jl/blob/next-speed/src/dBdt.jl#L172). Yet some species that go under the threshold for extinction (`10.0*eps()`) sometimes reappear. Do you have any suggestion?

Code for this figure:

~~~ julia
using BioEnergeticFoodWebs
using Plots

srand(12345)

S = 25
A = nichemodel(S, 0.2)
parameters = model_parameters(A)
bm = rand(Float64, S).*(1/S)

@time sim = simulate(parameters, bm; stop=500, use=:stiff)

minimum(sim[:B])

pl2 = vec(sum(sim[:B].>0.0, 2)) |> x -> plot(sim[:t], x, leg=false)
pl1 = plot(sim[:t], sim[:B], leg=false, c=:teal)
plot(pl1, pl2)
~~~

ChrisRackauckas commented:
The continuous callback interp_points are used to try to find the first time point it goes below the threshold. I'm curious as to why a DiscreteCallback was chosen.

tpoisot commented:
My understanding of the doc ([this part in particular](http://docs.juliadiffeq.org/latest/features/callback_functions.html#ContinuousCallbacks-1)) is that the condition needs to return `0`, and discrete callbacks are called when something returns `true` -- so I thought the discrete callback would trigger whenever it detects a value of `integrator.u` below some threshold.

ChrisRackauckas commented:
>My understanding of the doc (this part in particular) is that the condition needs to return 0, and discrete callbacks are called when something returns true -- so I thought the discrete callback would trigger whenever it detects a value of integrator.u below some threshold.

Yeah but that's only checking the endpoints. The issue is that within an interval you can dip below 0 and then come back above it. If you only check the end points you'll not catch that it went extinct. But at the same time, if you plot with the interpolation it'll fill in the values within the timestep and show that it actually went extinct.

The continuous callback protects against this by checking at `interp_points` evenly spaced points throughout the interval to check for a zero crossing, and then it tries to find the exact time point of the first zero crossing. If the dip is quick of course it can't catch it (and catching it would be almost impossible, at least without doing some fancy local minima search), but it protects against this issue.

tpoisot committed:
example figure

ChrisRackauckas commented:
I'm pretty much developing only on v0.7 now, so I'm going to investigate it there and try to narrow down what's happening. If you need it before JuliaCon and the v0.7 release, let me know and I can backport it.

tpoisot committed:
:hand: restore ContinuousCallback

tpoisot committed:
:tada: fix extinction callback -- species don't come back from extinction

tpoisot committed:
:hand:

tpoisot committed:
clean dBdt
